### ISSUE TITLE
[java] UnusedAssignment false positive when reporting unused variables

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#3114]
**Affects PMD Version:** 
6.31.0

**Rule:**
[UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment)
Property reportUnusedVariables=true

**Description:**
False positive when declared variables are initialized and referenced in multi-line code-blocks surrounded by curly braces.

**Code Sample demonstrating the issue:**

```java
import java.io.IOException;
import static org.objectweb.asm.Opcodes.*;
import org.objectweb.asm.ClassWriter; 
import org.objectweb.asm.FieldVisitor; 
import org.objectweb.asm.MethodVisitor;

public class Dumper {
    private static final String VERSION = "version";
    private static final String CLASS = "Dumper";

    public static byte[] dump() throws IOException {
        // ClassWriter is a class visitor that generates the code for the class
        ClassWriter cw = new ClassWriter(0);
        // Start creating the class.
        cw.visit(V11, ACC_PUBLIC + ACC_SUPER, CLASS, null, "java/lang/Object", null);
        FieldVisitor fv; // false positive
        MethodVisitor mv; // false positive
        {
            // version field
            fv = cw.visitField(ACC_PRIVATE, VERSION, "I", null, null);
            fv.visitEnd();
        }
        {
            // Implementing the constructor
            mv = cw.visitMethod(ACC_PUBLIC, "<init>", "()V", null, null);
            mv.visitCode();
            mv.visitVarInsn(ALOAD, 0);
            mv.visitMethodInsn(INVOKESPECIAL, "java/lang/Object", "<init>", "()V", false);
            mv.visitInsn(RETURN);
            mv.visitMaxs(1, 1);
            mv.visitEnd();
        }
        cw.visitEnd();

        return cw.toByteArray();
    }
}
```

Second example:
```java
 int start;
     if (isOdd(firstPartitionSize))       start = firstPartitionSize;
     else start =                         firstPartitionSize + 1;
```

Third example
```java
private boolean isPair(int number) {
    // Unlike SoftWorker's isPair(int) method,
    // this one contains a little bit more code not necessarily
    // related to pair number detection. It's here to detect if
    // a long method can be inlined
    String stringNumber = String.valueOf(number);
    int divRest;
    if (stringNumber.length() == 3)
      divRest = number % 10;
    else if (stringNumber.length() == 4)
      divRest = number % 100;
    else if (stringNumber.length() == 5)
      divRest = number % 1000;
    else if (stringNumber.length() == 6)
      divRest = number % 1000;
    else
      divRest = number % 10_000;
    if (divRest == 0)
      return true;
    return number % 2 == 0;
  }
```

**Expected outcome:**

No violation reported.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by Fernal73** [comment]
@oowekyala This is labeled with false-negative as well.

**Comment by oowekyala** [comment]
Yes, it's on purpose

**Comment by Fernal73** [comment]
> Yes, it's on purpose

It's confusing.

**Comment by Fernal73** [comment]
Updated issue with another code sample in https://github.com/pmd/pmd/issues/3114#issue-807788878
